src/coreclr/nativeaot/System.Private.CoreLib/src/System/CrashInfo.cs

```diff
@@ -0,0 +1,389 @@
+
+            AggregateException? aggregate = exception as AggregateException;
+            if (aggregate is not null || exception.InnerException is not null)
+            {
+                success = WriteBlock("inner", '[', ']', () =>
+                {
+                    if (aggregate is not null)
+                    {
+                        foreach (Exception ex in aggregate.InnerExceptions)
+                        {
+                            if (!WriteBlock(null, '{', '}', () => WriteException(ex, maxMessageSize, maxNumberStackFrames, maxMethodNameSize)))
+                                return false;
+                        }
+                    }
+                    else
+                    {
+                        if (!WriteBlock(null, '{', '}', () => WriteException(exception.InnerException, maxMessageSize, maxNumberStackFrames, maxMethodNameSize)))
+                            return false;
+                    }
+                    return true;
+                });
+
+                if (!success)
+                    return false;
+            }
+            return true;
+        }
+
+        /// <summary>
+        /// Write an exception stack frame to the triage buffer
+        /// </summary>
+        /// <param name="frame">the stack frame instance to write</param>
+        /// <param name="maxNameSize">limits the size of the frame type name</param>
+        /// <returns>true - success, false - out of triage buffer space</returns>
+        private bool WriteStackFrame(StackFrame frame, int maxNameSize)
+        {
+            return WriteBlock(null, '{', '}', () =>
+            {
+                nint ip = frame.GetNativeIPAddress();
+                if (!WriteHexValue("ip", (ulong)ip))
+                    return false;
+
+                if (!WriteHexValue("offset", frame.GetNativeOffset()))
+                    return false;
+
+                string method = DeveloperExperience.GetMethodName(ip, out IntPtr _);
+                if (method != null)
+                {
+                    if (!WriteValue("name", method, maxNameSize))
+                        return false;
+                }
+                return true;
+            });
+        }
+
+        private bool WriteHexValue(ReadOnlySpan<char> key, ulong value) => WriteValue(key, $"0x{value:X}");
+
+        private bool WriteHexValue(ReadOnlySpan<char> key, int value) => WriteValue(key, $"0x{value:X}");
+
+        private bool WriteValue(ReadOnlySpan<char> key, int value) => WriteValue(key, value.ToString());
+
+        private bool WriteValue(ReadOnlySpan<char> key, string value, int max = int.MaxValue)
+        {
+            // Escape the special JSON characters. It is done here and not at any lower level
+            // because this function is the only one that could be passed a string that needs
+            // escape and the lower level functions actually need to write double quotes.
+            StringBuilder sb = new StringBuilder();
```

Comment:
Allocating a StringBuilder to every arbitrary string we need to write is a lot of extra allocations. It will work, but it goes against the idea of keeping the allocations reasonably minimal in this code.

I think the escaping should be done in the lower-level code, without extra allocations.

Reply:
E.g. if you assume that the average string written through this method is going to be 40 characters long, it is 300 bytes of worth of allocations per string.

Reply:
I know this is really ugly but the lower-level code needs to be able to write a double quote and never enumerates each individual char. It uses Encoding.UTF8.GetBytes to copy the converted bytes to the triage buffer.  Probably have to pass a flag down to escape/not-escape to CrashInfo.WriteChars. I'll think about it and come up with something.

Reply:
Another approach would be not to escape and assume there are never control chars, " or \ in exception messages or type names.

Reply:
`\` in exception messages is very common. For example, try `File.Open(@"c:\doesnotexist.txt", FileMode.Open);`